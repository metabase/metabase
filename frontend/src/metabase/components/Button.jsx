import React from "react";
import PropTypes from "prop-types";

import Icon from "metabase/components/Icon.jsx";

import cx from "classnames";

import _ from "underscore";

const BUTTON_VARIANTS = [
  "small",
  "medium",
  "large",
  "primary",
  "warning",
  "cancel",
  "success",
  "purple",
  "borderless",
  "onlyIcon",
];

const Button = ({
  className,
  icon,
  iconRight,
  iconSize,
  children,
  ...props
}) => {
  let variantClasses = BUTTON_VARIANTS.filter(variant => props[variant]).map(
    variant => "Button--" + variant,
  );

  return (
    <button
      {..._.omit(props, ...BUTTON_VARIANTS)}
      className={cx("Button", className, variantClasses)}
    >
      <div className="flex layout-centered">
        {icon && (
          <Icon
            name={icon}
            size={iconSize ? iconSize : 14}
            className={cx({ mr1: !props.onlyIcon })}
          />
        )}
        <div>{children}</div>
        {iconRight && (
          <Icon
            name={iconRight}
            size={iconSize ? iconSize : 14}
            className={cx({ ml1: !props.onlyIcon })}
          />
        )}
      </div>
    </button>
  );
};

Button.propTypes = {
  className: PropTypes.string,
  icon: PropTypes.string,
  iconSize: PropTypes.number,
  children: PropTypes.any,

  small: PropTypes.bool,
  medium: PropTypes.bool,
  large: PropTypes.bool,

  primary: PropTypes.bool,
  warning: PropTypes.bool,
  cancel: PropTypes.bool,
  purple: PropTypes.bool,

  borderless: PropTypes.bool,
  onlyIcon: PropTypes.bool,
};

export default Button;
