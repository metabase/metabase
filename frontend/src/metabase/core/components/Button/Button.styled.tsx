import styled from "@emotion/styled";
import { css } from "@emotion/react";

export const ButtonRoot = styled.button`
  transition: all 200ms linear;
  flex-shrink: 0;
`;

export interface ButtonContentProps {
  iconVertical?: boolean;
}

export const ButtonContent = styled.div<ButtonContentProps>`
  display: flex;
  flex-direction: ${props => (props.iconVertical ? "column" : "row")};
  align-items: center;
  justify-content: center;
  min-width: ${props => (props.iconVertical ? "60px" : "")};
`;

export interface ButtonTextContainerProps {
  iconVertical: boolean;
  hasIcon: boolean;
  hasRightIcon: boolean;
}

const verticalTopIconCSS = css`
  margin-top: 0.5rem;
`;

const verticalBottomIconCSS = css`
  margin-bottom: 0.5rem;
`;

const horizontalLeftIconCSS = css`
  margin-left: 0.5rem;
`;

const horizontalRightIconCSS = css`
  margin-right: 0.5rem;
`;

export const ButtonTextContainer = styled.div<ButtonTextContainerProps>`
  ${props => {
    if (props.hasIcon) {
      return props.iconVertical ? verticalTopIconCSS : horizontalLeftIconCSS;
    }
  }}

  ${props => {
    if (props.hasRightIcon) {
      return props.iconVertical
        ? verticalBottomIconCSS
        : horizontalRightIconCSS;
    }
  }}
`;
