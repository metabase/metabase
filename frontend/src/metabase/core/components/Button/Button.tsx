import cx from "classnames";
import React, {
  ButtonHTMLAttributes,
  forwardRef,
  ReactNode,
  Ref,
  ElementType,
} from "react";
import styled from "@emotion/styled";
import { color, space } from "styled-system";
import _ from "underscore";
import Icon from "metabase/components/Icon";
import {
  ButtonContent,
  ButtonRoot,
  ButtonTextContainer,
} from "./Button.styled";

const BUTTON_VARIANTS = [
  "small",
  "medium",
  "large",
  "round",
  "primary",
  "danger",
  "warning",
  "cancel",
  "success",
  "purple",
  "white",
  "borderless",
  "onlyIcon",
  "fullWidth",
] as const;

interface Props extends ButtonHTMLAttributes<HTMLButtonElement> {
  as?: ElementType;
  className?: string;

  icon?: string;
  iconSize?: number;
  iconColor?: string;
  iconRight?: string;
  iconVertical?: boolean;
  labelBreakpoint?: string;
  children?: ReactNode;

  small?: boolean;
  medium?: boolean;
  large?: boolean;

  primary?: boolean;
  success?: boolean;
  danger?: boolean;
  warning?: boolean;
  cancel?: boolean;
  white?: boolean;
  purple?: boolean;

  disabled?: boolean;
  round?: boolean;
  borderless?: boolean;
  onlyIcon?: boolean;
  fullWidth?: boolean;
}

const BaseButton = forwardRef(function BaseButton(
  {
    as,
    className,
    icon,
    iconRight,
    iconSize,
    iconColor,
    iconVertical = false,
    labelBreakpoint,
    children,
    ...props
  }: Props,
  ref: Ref<HTMLButtonElement>,
) {
  const variantClasses = BUTTON_VARIANTS.filter(variant => props[variant]).map(
    variant => "Button--" + variant,
  );

  return (
    <ButtonRoot
      as={as}
      {..._.omit(props, ...BUTTON_VARIANTS)}
      className={cx("Button", className, variantClasses, {
        p1: !children,
      })}
      ref={ref}
    >
      <ButtonContent iconVertical={iconVertical}>
        {icon && (
          <Icon color={iconColor} name={icon} size={iconSize ? iconSize : 14} />
        )}
        {children && (
          <ButtonTextContainer
            hasIcon={!!icon}
            hasRightIcon={!!iconRight}
            iconVertical={iconVertical}
            className={cx({
              [`hide ${labelBreakpoint}-show`]: !!labelBreakpoint,
            })}
          >
            {children}
          </ButtonTextContainer>
        )}
        {iconRight && (
          <Icon
            color={iconColor}
            name={iconRight}
            size={iconSize ? iconSize : 14}
          />
        )}
      </ButtonContent>
    </ButtonRoot>
  );
});

const Button = styled(BaseButton)`
  ${color};
  ${space};
`;

Button.displayName = "Button";

export default Object.assign(Button, {
  Root: ButtonRoot,
  Content: ButtonContent,
  TextContainer: ButtonTextContainer,
});
