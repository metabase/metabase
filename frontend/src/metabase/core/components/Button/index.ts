export { default } from "./Button";
export type { ButtonProps } from "./Button";
