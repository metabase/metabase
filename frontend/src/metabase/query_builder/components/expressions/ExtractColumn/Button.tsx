import { Text, Box } from "metabase/ui";

import styles from "./Button.module.css";

export function Button({
  title,
  example,
  onClick,
}: {
  title: string;
  example: string;
  onClick: () => void;
}) {
  return (
    <Box
      component="button"
      type="button"
      p="sm"
      className={styles.button}
      onClick={onClick}
    >
      <Text color="text-dark" className={styles.content} weight="bold" p={0}>
        {title}
      </Text>
      <Text color="text-light" size="sm" className={styles.content}>
        {example}
      </Text>
    </Box>
  );
}
