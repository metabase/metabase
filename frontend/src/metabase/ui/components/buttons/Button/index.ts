export { Button } from "@mantine/core";
export type { ButtonProps, ButtonGroupProps } from "@mantine/core";
export { getButtonOverrides } from "./Button.styled";
