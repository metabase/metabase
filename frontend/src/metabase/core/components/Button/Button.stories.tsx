import type { ComponentStory } from "@storybook/react";
import Button from "./Button";

export default {
  title: "Core/Button",
  component: Button,
};

const Template: ComponentStory<typeof Button> = args => {
  return <Button {...args} />;
};

export const Default = Template.bind({});
Default.args = {
  children: "Default",
};

export const Primary = Template.bind({});
Primary.args = {
  primary: true,
  children: "Primary",
};

export const WithIcon = Template.bind({});
WithIcon.args = {
  icon: "chevrondown",
};

export const OnlyText = Template.bind({});
OnlyText.args = {
  onlyText: true,
  children: "Click Me",
};
