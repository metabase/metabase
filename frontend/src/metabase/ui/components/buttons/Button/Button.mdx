import { Fragment } from "react";
import { Canvas, Story, Meta } from "@storybook/blocks";
import { Icon } from "metabase/ui";
import { Group, Stack } from "metabase/ui";
import { Button } from "./";
import * as ButtonStories from "./Button.stories";

<Meta of={ButtonStories} />

# Button

Our themed wrapper around [Mantine Button](https://v6.mantine.dev/core/button/).

## When to use Button

Use button in the following cases:

- Buttons are widely used as actions that users can take. They are typically placed on the UI such as dialogs, forms, toolbars, config pages/panels, headers etc.
- Use primary button for primarily intended actions.
- Use icon in primary button for improved visual affordance.
- Primary buttons could be colorized with our brand colors for differentiating actions and states.

Not to use:

- Avoid using multiple primary buttons on a form section, or a dialog for competing primarily intended actions.
- If there is no primarily intended action for a form, use default (non-primary) button or other variations. That is, you donâ€™t have to have a primary button on the form or dialog.
- Not to use links to replace subtle buttons. Links are meant to navigate to another page. Subtle buttons share the characteristics of button and are mostly used in compact or inline situations.

## Docs

- [Figma File](https://www.figma.com/file/Ey1rOyIxRHpmRvE9XrGyop/Buttons-%2F-Button?type=design&node-id=1-96&mode=design&t=yaNljw178EFJeU7k-0)
- [Mantine Button Docs](https://v6.mantine.dev/core/button/)

## Caveats

- Use verb, or verb+noun for button labels. E.g., Save, Filter, Summarize, Convert this question to SQL.
- Use sentence casing for long button labels.

## Usage guidelines

- Although button labels can be as long as you wish, keep it simple and concise.
- In general, use sentence casing for menu item labels.
- Use icon+button label when applicable for better visual affordance. E.g., + New

## Examples

<Canvas>
  <Story of={ButtonStories.Default} />
</Canvas>

### Button.Group

<Canvas>
  <Story of={ButtonStories.ButtonGroup} />
</Canvas>

### Default size

<Canvas>
  <Story of={ButtonStories.DefaultSize} />
</Canvas>

#### Custom color

<Canvas>
  <Story of={ButtonStories.DefaultSizeCustomColor} />
</Canvas>

#### Disabled state

<Canvas>
  <Story of={ButtonStories.DefaultSizeDisabled} />
</Canvas>

#### Loading state

<Canvas>
  <Story of={ButtonStories.DefaultSizeLoading} />
</Canvas>

### Default size & full width

<Canvas>
  <Story of={ButtonStories.DefaultSizeFullWidth} />
</Canvas>

#### Disabled state

<Canvas>
  <Story of={ButtonStories.DefaultSizeFullWidthDisabled} />
</Canvas>

#### Loading state

<Canvas>
  <Story of={ButtonStories.DefaultSizeFullWidthLoading} />
</Canvas>

### Compact size

<Canvas>
  <Story of={ButtonStories.CompactSize} />
</Canvas>

#### Custom color

<Canvas>
  <Story of={ButtonStories.CompactSizeCustomColor} />
</Canvas>

#### Disabled state

<Canvas>
  <Story of={ButtonStories.CompactSizeDisabled} />
</Canvas>

#### Loading state

<Canvas>
  <Story of={ButtonStories.CompactSizeLoading} />
</Canvas>

### Compact size & full width

<Canvas>
  <Story of={ButtonStories.CompactSizeFullWidth} />
</Canvas>

#### Disabled state

<Canvas>
  <Story of={ButtonStories.CompactSizeFullWidthDisabled} />
</Canvas>

#### Loading state

<Canvas>
  <Story of={ButtonStories.CompactSizeFullWidthLoading} />
</Canvas>
