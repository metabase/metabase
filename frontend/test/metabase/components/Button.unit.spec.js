import React from "react";
import renderer from "react-test-renderer";

import { render } from "enzyme";

import Button from "metabase/components/Button";

describe("Button", () => {
  it("should render correctly", () => {
    const tree = renderer.create(<Button>Clickity click</Button>).toJSON();

    expect(tree).toMatchSnapshot();
  });
  it("should render correctly with an icon", () => {
    const tree = renderer
      .create(<Button icon="star">Clickity click</Button>)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it("should render a primary button given the primary prop", () => {
    const button = render(<Button primary>Clickity click</Button>);

    expect(button.hasClass("Button--primary")).toBe(true);
  });
});
