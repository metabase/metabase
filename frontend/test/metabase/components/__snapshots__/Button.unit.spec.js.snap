// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should render correctly 1`] = `
<button
  className="Button sc-bdVaJa bDWFJH flex-no-shrink "
>
  <div
    className="flex layout-centered"
    style={null}
  >
    <div
      className=""
    >
      Clickity click
    </div>
  </div>
</button>
`;

exports[`Button should render correctly with an icon 1`] = `
<button
  className="Button sc-bdVaJa bDWFJH flex-no-shrink "
>
  <div
    className="flex layout-centered"
    style={null}
  >
    <svg
      className="Icon Icon-star Icon-cxuQhR kTAgZA"
      fill="currentcolor"
      height={14}
      viewBox="0 0 32 32"
      width={14}
    >
      <path
        d="M16 0 L21 11 L32 12 L23 19 L26 31 L16 25 L6 31 L9 19 L0 12 L11 11"
      />
    </svg>
    <div
      className="ml1"
    >
      Clickity click
    </div>
  </div>
</button>
`;
